controllers/managedfusiondeployment_controller.go

```diff
@@ -0,0 +1,795 @@
+			deployment := &deployerCSVDeployments[key]
+			if deployment.Name == "managed-fusion-controller-manager" {
+				deployerCSVDeployment = deployment
+			}
+		}
+
+		deployerCSVContainers := deployerCSVDeployment.Spec.Template.Spec.Containers
+		var kubeRbacImage string
+		for key := range deployerCSVContainers {
+			container := deployerCSVContainers[key]
+			if container.Name == "kube-rbac-proxy" {
+				kubeRbacImage = container.Image
+			}
+		}
+
+		prometheusContainers := desired.Spec.Containers
+		for key := range prometheusContainers {
+			container := &prometheusContainers[key]
+			if container.Name == "kube-rbac-proxy" {
+				container.Image = kubeRbacImage
+			}
+		}
+
+		clusterVersion := &configv1.ClusterVersion{}
+		clusterVersion.Name = "version"
+		if err := r.get(clusterVersion); err != nil {
+			return err
+		}
+
+		r.prometheus.Spec = desired.Spec
+		r.prometheus.Spec.ExternalLabels["clusterId"] = string(clusterVersion.Spec.ClusterID)
+		r.prometheus.Spec.Alerting.Alertmanagers[0].Namespace = r.Namespace
+		r.prometheus.Spec.AdditionalAlertRelabelConfigs = &corev1.SecretKeySelector{
+			LocalObjectReference: corev1.LocalObjectReference{
+				Name: alertRelabelConfigSecretName,
+			},
+			Key: alertRelabelConfigSecretKey,
+		}
+
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("Failed to update Prometheus: %v", err)
+	}
+
+	return nil
+}
+
+func (r *ManagedFusionDeployment) reconcileAlertmanager() error {
+	r.Log.Info("Reconciling Alertmanager")
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.alertmanager, func() error {
+		if err := r.own(r.alertmanager); err != nil {
+			return err
+		}
+
+		desired := templates.AlertmanagerTemplate.DeepCopy()
+		desired.Spec.AlertmanagerConfigSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				monLabelKey: monLabelValue,
+			},
+		}
+		r.alertmanager.Spec = desired.Spec
+		utils.AddLabel(r.alertmanager, monLabelKey, monLabelValue)
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *ManagedFusionDeployment) reconcileAlertmanagerSecret() error {
+	r.Log.Info("Reconciling AlertmanagerSecret")
+	managedFusionSecretData := r.managedFusionSecret.Data
+	managedFusionDeploymentPagerDutyConfig, exist := managedFusionSecretData["pager_duty_config"]
+	if !exist {
+		return fmt.Errorf("pager_duty_config key not found in managed-fusion-agent-config secret")
+	}
+	if err := yaml.Unmarshal(managedFusionDeploymentPagerDutyConfig, r.pagerDutyConfigData); err != nil {
+		return err
+	}
+	if r.pagerDutyConfigData.ServiceKey == "" {
+		return fmt.Errorf("managed-fusion-agent-config secret contains an empty pagerDutyKey entry")
+	}
+
+	managedFusionDeploymentSMTPConfig, exist := managedFusionSecretData["smtp_config"]
+	if !exist {
+		return fmt.Errorf("smtp_config key not found in managed-fusion-agent-config secret")
+	}
+	if err := yaml.Unmarshal(managedFusionDeploymentSMTPConfig, r.smtpConfigData); err != nil {
+		return err
+	}
```

Comment:
This should be the first thing we are doing in the reconciliation flow (reconciler function)! Without a structurally valid config reconciling does not make sense. The assumption here should be that r. smtpSpec is already populated

Reply:
Updated in latest commit